### PR TITLE
[spmd_types] MoE sparse mesh transitions

### PR DESCRIPTION
**Pull request by pianpwk**
`set_current_spmd_mesh(sparse_mesh)` transitions for MoE region, plus mesh reinterpret at transition points, 1 assert_type re-annotation, 1 mutate_type for grouped_mm

Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.12.0) (oldest at bottom):
* #3740
* #3690
* #3673
* #3655
* #3741
* #3664
* __->__ #3654

### CODE REVIEW

File: torchtitan/models/common/moe.py
Comment (tianyu-l) [inline torchtitan/models/common/moe.py:166]: man with spmd_types I wanted to get rid of this hacky `wire_mesh`, but now this PR is exploiting it, lol

**Review by tianyu-l — CHANGES_REQUESTED** [review]

File: torchtitan/models/common/moe.py
Comment (pianpwk) [reply torchtitan/models/common/moe.py:166]: so we still need this for full_dtensor, and since the transition point doesn't fall on module boundary, how do you prefer we access & set this

Comment (tianyu-l) [reply torchtitan/models/common/moe.py:166]: Is it only needed for full_dtensor? I'm fine with it until we finished migration, but do you still need this after migration is done and we are spmd_types only? I would like to push back on that. We can discuss.

Comment (pianpwk) [reply torchtitan/models/common/moe.py:166]: sorry, that comment was conflating 2 things. changed, but we still need to set sparse_mesh somewhere for spmd backend.

File: torchtitan/models/common/moe.py [hunk 1]
```diff
@@ -76,8 +79,15 @@ def _experts_forward(
...
             w3_EFD = self.w3_EFD
 
-        offsets_E = torch.cumsum(num_tokens_per_expert_E, dim=0, dtype=torch.int32)
+        if (
+            get_spmd_backend() == "spmd_types"
+            and spmd.is_type_checking()
+            and spmd_mesh_size("ep") == 1
+        ):
+            for axis in ("dp", "cp"):  # if no EP, convert to V for grouped_mm
+                spmd.mutate_type(num_tokens_per_expert_E, axis, src=spmd.P, dst=spmd.V)
```
Comment (tianyu-l) [inline torchtitan/models/common/moe.py]: trying to understand this better:
1. with EP this will automatically be spmd.V (sharded on num_experts dim), but when EP is off this should be spmd.P but it couldn't work with grouped mm? If so we need way more comments to explain.
2. why do you put this mutate on `num_tokens_per_expert_E` not `offsets_E`?
3. could you remind me when do we choose `mutate` over `reinterpret`? I think this field doesn't need gradient anyway.
4. actually, without EP this can be a global region instead of a local region. How do you think we can do that? (a) if EP is off we don't annotate it with local_spmd=True in the module (b) lightweight solution could be scoop a global region from the local map module.

File: torchtitan/models/common/moe.py [hunk 2]
```diff
@@ -152,6 +163,10 @@ def parallelize(self, parallel_dims) -> None:
             ep_mesh=parallel_dims.get_optional_mesh("ep"),
             tp_mesh=parallel_dims.get_optional_mesh("tp"),
         )
+        self.token_dispatcher.sparse_mesh = parallel_dims.get_optional_mesh(
+            ["dp_replicate", "efsdp", "ep"],
+            include_singleton_axes=True,
+        )
```
Comment (tianyu-l) [inline torchtitan/models/common/moe.py]: I dislike how this is done today.

Right now `sparse_mesh` is an attribute of token_dispatcher, but it was not set at construction time. Instead, it's later assigned by this `parallelize` function, which is not any cleaner than using a global context.

There are two ways I could think of
1. we explicitly set two meshes upon entering fwdbwd https://github.com/pytorch/torchtitan/blob/main/torchtitan/distributed/utils.py#L386, one dense one sparse, and activate the dense one first by default, then activate the sparse one in MoE
2. when parallel_dims is constructed, store two meshes in global var; activate the dense one here https://github.com/pytorch/torchtitan/blob/main/torchtitan/distributed/utils.py#L386 (Notice that today you are using private variables to activate dense mesh, which should be improved anyway.) 

I think we only need minimal changes for 2.

File: torchtitan/models/common/token_dispatcher.py [hunk 3]
```diff
@@ -330,33 +350,52 @@ def dispatch(
...
-        ) = self._permute(
-            routed_input_RD,
-            num_global_tokens_per_local_expert_E,
-        )
+        with self.sparse_spmd_mesh():
+            if get_spmd_backend() == "spmd_types":
+                # re-annotate after no-typecheck
+                spmd.assert_type(
+                    num_global_tokens_per_local_expert_E,
+                    {"dp_replicate": spmd.V, "efsdp": spmd.V, "ep": spmd.S(0)},
```
Comment (tianyu-l) [inline torchtitan/models/common/token_dispatcher.py]: why it's V not P?

File: torchtitan/models/common/token_dispatcher.py [hunk 4]
```diff
@@ -330,33 +350,52 @@ def dispatch(
...
+                )
+                if spmd.is_type_checking():  # sparse mesh reinterpret
+                    routed_input_ND = spmd.reinterpret_mesh(
+                        routed_input_ND, spmd.current_mesh()
+                    )
```
Comment (tianyu-l) [inline torchtitan/models/common/token_dispatcher.py]: It looks nice to have this explicit `reinterpret_mesh`. Just wonder if it could be implicitly inferred from the `spmd.all_to_all` below? I forgot when can we infer.

File: torchtitan/models/common/token_dispatcher.py
Comment (tianyu-l) [inline torchtitan/models/common/token_dispatcher.py:647]: Is the plan to support these variants later?

File: torchtitan/models/common/token_dispatcher.py [hunk 5]
```diff
@@ -210,6 +220,16 @@ def wire_meshes(
...
 
+    @contextmanager
+    def sparse_spmd_mesh(self) -> Iterator[None]:
+        if self.ep_mesh is None or get_spmd_backend() != "spmd_types":
+            yield
+            return
+
+        assert self.sparse_mesh is not None, "spmd_types EP dispatch requires sparse_mesh"
+        with set_current_spmd_mesh(self.sparse_mesh):
+            yield
```
Comment (tianyu-l) [inline torchtitan/models/common/token_dispatcher.py]: similar, this should be replaced by global context

**Review by tianyu-l — CHANGES_REQUESTED** [review]

File: torchtitan/models/common/moe.py (hunk 1 above)
Comment (pianpwk) [reply torchtitan/models/common/moe.py]: > with EP this will automatically be spmd.V (sharded on num_experts dim), but when EP is off this should be spmd.P but it couldn't work with grouped mm? If so we need way more comments to explain.

adding a comment, grouped_mm would see `x:R, w1:V, offsets:P` in local SPMD typechecking, which is banned because spmd.P is not currently allowed to mix with spmd.V. This is too strict, DTensor contains valid Shard + Partial rules, so TODO is likely to relax this in spmd_types.

> why do you put this mutate on num_tokens_per_expert_E not offsets_E?

no real reason, either works

> could you remind me when do we choose mutate over reinterpret? I think this field doesn't need gradient anyway.

reinterpret should call an autograd function, so there is a corresponding BWD. reinterpret(P->anything) is also banned. mutate is when we want to brute-force change the type, I don't think there's BWD behavior. Since this doesn't need a gradient I think that's ok ?

> actually, without EP this can be a global region instead of a local region. How do you think we can do that? (a) if EP is off we don't annotate it with local_spmd=True in the module (b) lightweight solution could be scoop a global region from the local map module.

this should be doable, updating in moe_sharding.py

File: torchtitan/models/common/token_dispatcher.py
Comment (pianpwk) [reply torchtitan/models/common/token_dispatcher.py:647]: i haven't tried setup for these yet...

Comment (tianyu-l) [reply torchtitan/models/common/token_dispatcher.py:647]: We don't have to support explicit type checking into them in this PR. Just want to make sure that if we turn off type checking, it is still correct -- as long as you didn't introduce explicit redistribute for the default dispatcher, I think others would just follow the correctness of this PR.

File: torchtitan/models/common/token_dispatcher.py (hunk 3 above)
Comment (pianpwk) [reply torchtitan/models/common/token_dispatcher.py]: ah same as the "P doesn't mix" issue above

Comment (tianyu-l) [reply torchtitan/models/common/token_dispatcher.py]: I think we should mark them P here, and mutate everything later. The reason is simple, after you fix the local spmd sharding prop, here it should still be P.

File: torchtitan/models/common/moe.py [hunk 6]
```diff
@@ -127,9 +141,10 @@ def forward(
             topk_expert_ids_TK,
             num_local_tokens_per_expert_E,
         )
-        routed_output_RD = self._experts_forward(
-            routed_input_RD, num_global_tokens_per_local_expert_e
-        )
+        with self.token_dispatcher.sparse_spmd_mesh():
```
Comment (tianyu-l) [inline torchtitan/models/common/moe.py]: I believe you plan to update this

File: torchtitan/models/common/token_dispatcher.py [hunk 7]
```diff
@@ -64,6 +73,10 @@ def wire_meshes(
         """No-op for the EP=1 dispatcher. Subclasses override."""
         del ep_mesh, tp_mesh
 
+    @contextmanager
+    def sparse_spmd_mesh(self) -> Iterator[None]:
+        yield
```
Comment (tianyu-l) [inline torchtitan/models/common/token_dispatcher.py]: and this

File: torchtitan/models/common/token_dispatcher.py [hunk 8]
```diff
@@ -210,6 +223,15 @@ def wire_meshes(
             self.sp_size = tp_mesh.size()
             self.sp_rank = tp_mesh._sym_get_coordinate(0)
 
+    @contextmanager
+    def sparse_spmd_mesh(self) -> Iterator[None]:
```
Comment (tianyu-l) [inline torchtitan/models/common/token_dispatcher.py]: and this

File: torchtitan/models/common/token_dispatcher.py [hunk 9]
```diff
@@ -470,19 +518,35 @@ def combine(
...
+                routed_output_RD = spmd.all_to_all(
+                    routed_output_RD,
+                    current_spmd_mesh().get_group("ep"),
+                    src=spmd.S(0),
+                    dst=spmd.S(0),
+                    output_split_sizes=metadata.input_splits,
+                    input_split_sizes=metadata.output_splits,
+                )
+            else:
+                routed_output_RD = all_to_all_single(
```
Comment (tianyu-l) [inline torchtitan/models/common/token_dispatcher.py]: hmm, we should be able to do spmd.all_to_all anyway?

File: torchtitan/models/common/token_dispatcher.py [hunk 10]
```diff
@@ -330,33 +352,59 @@ def dispatch(
...
+                )
+                if spmd.is_type_checking():  # sparse mesh reinterpret
+                    routed_input_ND = spmd.reinterpret_mesh(
+                        routed_input_ND, spmd.current_mesh()
+                    )
+                routed_input_RD = spmd.all_to_all(
+                    routed_input_ND,
+                    current_spmd_mesh().get_group("ep"),
+                    src=spmd.S(0),
+                    dst=spmd.S(0),
```
Comment (tianyu-l) [inline torchtitan/models/common/token_dispatcher.py]: spmd.V?

File: torchtitan/models/common/token_dispatcher.py [hunk 11]
```diff
@@ -330,33 +352,59 @@ def dispatch(
...
+                # Convert to V for _permute, otherwise typechecker errors with P + V.
+                for axis in ("dp_replicate", "efsdp"):
+                    spmd.mutate_type(
+                        num_global_tokens_per_local_expert_E,
+                        axis,
+                        src=spmd.P,
+                        dst=spmd.V,
+                    )
+            else:
+                routed_input_RD = all_to_all_single(
```
Comment (tianyu-l) [inline torchtitan/models/common/token_dispatcher.py]: should call spmd.all_to_all

File: torchtitan/models/common/token_dispatcher.py [hunk 12]
```diff
@@ -330,33 +352,59 @@ def dispatch(
...
+                    routed_input_ND,
+                    current_spmd_mesh().get_group("ep"),
+                    src=spmd.S(0),
+                    dst=spmd.S(0),
+                    output_split_sizes=output_splits_list,
+                    input_split_sizes=input_splits_list,
+                )
+                # Convert to V for _permute, otherwise typechecker errors with P + V.
+                for axis in ("dp_replicate", "efsdp"):
+                    spmd.mutate_type(
```
Comment (tianyu-l) [inline torchtitan/models/common/token_dispatcher.py]: mutate only before the op with limitation

File: torchtitan/distributed/spmd_types.py [hunk 13]
```diff
@@ -97,6 +123,17 @@ def set_current_spmd_mesh(mesh: DeviceMesh | None) -> Iterator[None]:
             assert popped is mesh
 
 
+@contextlib.contextmanager
+def set_sparse_mesh() -> Iterator[None]:
```
Comment (tianyu-l) [inline torchtitan/distributed/spmd_types.py]:
```suggestion
def maybe_set_sparse_mesh() -> Iterator[None]:
```

File: torchtitan/models/common/token_dispatcher.py [hunk 14]
```diff
@@ -302,7 +308,7 @@ def dispatch(
         ) = self._local_reorder(x_TD, topk_scores_TK, topk_expert_ids_TK)
 
         # generate the input splits and output splits for all-to-all
-        with torch.no_grad():
+        with torch.no_grad(), spmd.no_typecheck():
```
Comment (tianyu-l) [inline torchtitan/models/common/token_dispatcher.py]: Do we need `no_typecheck` here? What if we switch the `all_to_all_single` below to `spmd.all_to_all`? Is it that we need to `reinterpret(P, V)` before all-to-all and the ep_mesh is not available in the dense mesh tensor?

File: torchtitan/models/common/token_dispatcher.py [hunk 15]
```diff
@@ -316,6 +322,12 @@ def dispatch(
                     num_global_tokens_per_local_expert_E
                 )
             )
+            if get_spmd_backend() == "spmd_types":
+                # re-annotate after no-typecheck
+                spmd.assert_type(
+                    num_global_tokens_per_local_expert_E,
+                    {"dp_replicate": spmd.P, "efsdp": spmd.P, "ep": spmd.S(0)},
+                )
```
Comment (tianyu-l) [inline torchtitan/models/common/token_dispatcher.py]: oh, it seems you are using `no_typecheck`+`assert_type` to accomplish a `reinterpret_mesh`

File: torchtitan/models/common/token_dispatcher.py [hunk 16]
```diff
@@ -302,7 +308,7 @@ def dispatch(
         ) = self._local_reorder(x_TD, topk_scores_TK, topk_expert_ids_TK)
 
         # generate the input splits and output splits for all-to-all
-        with torch.no_grad():
+        with torch.no_grad(), spmd.no_typecheck():
             num_global_tokens_per_local_expert_E = all_to_all_single(
                 num_local_tokens_per_expert_E,
```
Comment (tianyu-l) [inline torchtitan/models/common/token_dispatcher.py]: please `assert_type` on the input of `no_typecheck` region. It feels `P` to me on all axes.

File: torchtitan/models/common/token_dispatcher.py (hunk 15 above)
Comment (tianyu-l) [reply torchtitan/models/common/token_dispatcher.py]: I don't think it's an S(0), maybe just V?
It looks like ` (e0,r0), (e1,r0), ..., (e0,r1), (e1,r1), ...  (rank-major)`

File: torchtitan/models/common/token_dispatcher.py [hunk 17]
```diff
@@ -387,6 +410,17 @@ def _permute(
...
+        if get_spmd_backend() == "spmd_types":
+            # Counts below become local segment lengths/offsets, and the expert
+            # offsets passed to grouped_mm after this function returns.
+            for axis in ("dp_replicate", "efsdp"):
+                spmd.mutate_type(
+                    num_global_tokens_per_local_expert_E,
+                    axis,
+                    src=spmd.P,
+                    dst=spmd.V,
+                )
```
Comment (tianyu-l) [inline torchtitan/models/common/token_dispatcher.py]: I apologize, I now think this tensor should never be spmd.P, they should just be born to be spmd.V. Can we put them back above where we do
```
spmd.assert_type(
    num_global_tokens_per_local_expert_E,
    {"dp_replicate": spmd.V, "efsdp": spmd.V, "ep": spmd.V},
)
```

**Review by tianyu-l — COMMENTED** [review]
last comments, I promise

File: torchtitan/models/common/token_dispatcher.py [hunk 18]
```diff
@@ -301,62 +307,112 @@ def dispatch(
...
+                else self.ep_mesh.get_group()
             )
-            # Need to wait explicitly because it is used by a triton kernel later
-            # which doesn't realize that AsyncCollectiveTensor needs unwrapping
-            num_global_tokens_per_local_expert_E = (
-                torch.ops._c10d_functional.wait_tensor(
-                    num_global_tokens_per_local_expert_E
+            if use_spmd_types:
+                num_local_tokens_per_expert_E = spmd.reinterpret_mesh(
+                    num_local_tokens_per_expert_E, spmd.current_mesh()
```
Comment (tianyu-l) [inline torchtitan/models/common/token_dispatcher.py:325]: is this no-op in erasure mode?

File: torchtitan/models/common/token_dispatcher.py [hunk 19]
```diff
@@ -301,62 +307,112 @@ def dispatch(
...
                 )
+                routed_input_ND = spmd.reinterpret_mesh(
+                    routed_input_ND, spmd.current_mesh()
+                )
+
+            with torch.no_grad():
+                if use_spmd_types:
+                    num_local_tokens_per_expert_EP_e = (
```
Comment (tianyu-l) [inline torchtitan/models/common/token_dispatcher.py]: Technically this naming is breaking convention, which is "a single letter means one tensor dim; among multiple letters there's no `_`".

https://github.com/pytorch/torchtitan/blob/main/torchtitan/models/common/moe.py#L21

If we don't want to introduce another letter just for "EP", this temporary violation sounds fine. But please comment explicitly.

File: torchtitan/models/common/token_dispatcher.py [hunk 20]
```diff
@@ -301,62 +307,112 @@ def dispatch(
...
+                        num_local_tokens_per_expert_E.view(ep_size, -1)
+                    )
+                    num_global_tokens_per_local_expert_EP_e = spmd.all_to_all(
```
Comment (tianyu-l) [inline torchtitan/models/common/token_dispatcher.py]: I guess the if-else is because DTensor doesn't support spmd collectives?

**Review by tianyu-l — APPROVED** [review]
